rholang/src/main/scala/coop/rchain/rholang/interpreter/PrettyPrinter.scala

```diff
@@ -0,0 +1,205 @@
+      // TODO: Figure out if we can prevent ScalaPB from generating
+      case ChannelInstance.Empty => "@Nil"
+    }
+
+  def buildString(t: GeneratedMessage): String =
+    t match {
+      case v: Var     => buildString(v)
+      case c: Channel => buildString(c)
+      case s: Send =>
+        buildString(s.chan.get) + {
+          if (s.persistent) "!!("
+          else "!("
+        } + {
+          ("" /: s.data.zipWithIndex) {
+            case (string, (datum, i)) =>
+              string + buildString(datum) + {
+                if (i != s.data.length - 1) ", "
+                else ""
+              }
+          }
+        } + ")"
+
+      case r: Receive =>
+        val (totalFree, bindsString) = ((freeShift, "") /: r.binds.zipWithIndex) {
+          case ((free, string), (bind, i)) =>
+            val (patternFree, bindString) =
+              PrettyPrinter(boundShift + free, 0)
+                .buildPattern(bind.patterns)
+            (free + patternFree, string + bindString + {
+              if (r.persistent) " <= " else " <- "
+            } + buildString(bind.source.get) + {
+              if (i != r.binds.length - 1) " ; "
+              else ""
+            })
+        }
+        "for( " + bindsString + " ) { " + PrettyPrinter(freeShift, boundShift + totalFree)
+          .buildString(r.body.get) + " }"
+
+      case e: Eval => "*" + buildString(e.channel.get)
+
+      case n: New =>
+        "new " + buildVariables(n.bindCount) + " in { " + PrettyPrinter(
+          freeShift,
+          boundShift + n.bindCount).buildString(n.p.get) + " }"
+
+      case e: Expr =>
+        buildString(e)
+
+      case m: Match =>
+        "match { " + buildString(m.target.get) + " } { " + {
+          for { (matchCase, i) <- m.cases.zipWithIndex } yield {
+            buildMatchCase(matchCase) + {
+              if (i != m.cases.length - 1) "; "
+              else ""
+            }
+          }
+        } + " }"
+
+      case g: GPrivate => g.id
+
+      case par: Par =>
+        if (isEmpty(par)) "Nil"
+        else {
+          val list =
+            List(par.sends, par.receives, par.evals, par.news, par.exprs, par.matches, par.ids)
+          ((false, "") /: list) {
+            case ((prevNonEmpty, string), items) =>
+              if (items.nonEmpty) {
+                (true, string + { if (prevNonEmpty) " | " else "" } + ("" /: items.zipWithIndex) {
+                  case (_string, (_par, index)) =>
+                    _string + buildString(_par) + {
+                      if (index != items.length - 1) " | " else ""
+                    }
+                })
+              } else (prevNonEmpty, string)
+          }
+        }._2
+
+      case _ => throw new Error("Attempt to print unknown GeneratedMessage type")
+    }
+
+  private def buildVariables(bindCount: Int): String =
+    buildSeq(
+      (0 until Math.min(MAX_NEW_VAR_COUNT, bindCount))
+        .map(i => GPrivate(s"x${boundShift + i}")))
```

Comment:
Why is this GPrivate? That just seems unnecessary.

Reply:
I think `i` should probably just map to a string. Not a language construct.